### PR TITLE
Q&A: Realized a number of newer features not in Staging yet

### PR DESCRIPTION
**Pull request by FlightfulKiwi**
Merged latest staging (w/o Ryan & Daniel's pending PRs) prior to this PR. Will to do another pull when both are merged.

- Updates sorting to 3 chars prior to Search Bar sorting
- Displays 4 questions on load
- Stores images in Cloudinary
- Creates thumbnails
- A few other features I've forgotten about
- BUGS: 
      - Posting Q&A's (mostly fine, minor kinks)
      - Re-rendering after a Post (might ignore altogether)
      - Closing Modal after submitting an Answer (but not a question)

### CODE REVIEW

File: client/src/components/QandA.jsx [hunk 1]
```diff
@@ -10,7 +10,8 @@ const QandA = ({ product_id, product_name }) => {
       <p id='QA-section-title'>Questions & Answers</p>
       <SearchBar setSearchText={setSearchText}/>
       <QuestionsList
-        product_id={40356}
+        //product_id={40356}
```
Comment (ryanaolea) [inline client/src/components/QandA.jsx]: still need?

File: client/src/components/QandA/QuestionsList.jsx [hunk 2]
```diff
@@ -14,19 +14,24 @@ class QuestionsList extends React.Component {
...
+  getAllQuestions() {
+    axios.get('http://127.0.0.1:3000/qa/questions', {params: { product_id: this.props.product_id, count: 50 }})
       .then(returnedQuestions => {
         this.setState({
           questions: returnedQuestions.data.results,
-          questionCounter: 2
+          questionCounter: 4
         });
       })
       .catch(err => console.error('Error receiving response (QuestionsList.jsx): ', err));
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList.jsx:28]: nice error messages! 😉

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 3]
```diff
@@ -1,16 +1,26 @@
...
 
+const cloudinary_name = 'flightfulkiwi';
+const cloudinary_url = 'https://api.cloudinary.com/v1_1/flightfulkiwi/image/upload';
+const cloudinary_preset = 'fec_preset';
+
 let title;
 let subtitle;
 let title_body;
+let buttonText;
+let imageErrorMsg = 'Sorry! Maximum number of images (5) exceeded.';
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:14]: this is an elegant way to handle this, nice work!

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 4]
```diff
@@ -20,109 +30,181 @@ const Modal = ({ setShowModal, usage, product_name, questionOrProduct_id }) => {
...
+    buttonText = 'Submit Answer';
   } else {
     title = 'Ask Your Question';
     subtitle = `About the ${product_name}:`;
     title_body = 'Your Question';
+    buttonText = 'Submit Question';
   }
 
-  const postModalInput = (endpoint, body) => {
-    // console.log('Modal Post Body: ', body);
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:34]: saving for testing?

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 5]
```diff
@@ -20,109 +30,181 @@ const Modal = ({ setShowModal, usage, product_name, questionOrProduct_id }) => {
...
+    const files = e.target.files;
+    const filesArray = [];
+    for (let i = 0; i < files.length; i++) {
+      filesArray.push(files[i]);
+    }
+
+    setImages(prevImages => [...filesArray, ...prevImages]);
+
+    const previewDIV = document.getElementById('QA-Modal-previews');
+    // while (previewDIV.firstChild) {
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:56]: mysterious while loop needed still ?

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 6]
```diff
@@ -20,109 +30,181 @@ const Modal = ({ setShowModal, usage, product_name, questionOrProduct_id }) => {
...
+    setPreviewSources([]);
+    setImages([]);
+  }
+
+  const handleSubmit = async (event) => {
     event.preventDefault();
+    const imageURLs = [];
+
+    // Uploads images to Cloudinary
+    for (let k = 0; k < images.length; k++) {
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:84]: smart!

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 7]
```diff
@@ -20,109 +30,181 @@ const Modal = ({ setShowModal, usage, product_name, questionOrProduct_id }) => {
...
+        imageURLs.unshift(result.data.url);
+      } catch(err) {
+        console.error( 'Error posting to Cloudinary: ', err );
+      }
+    }
 
     let endpoint;
     let body = {
       body: document.getElementById('QA-modal-body').value,
+      // body: event.target.modalBody.value,
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:101]: keeping for refactoring?

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx [hunk 8]
```diff
@@ -20,109 +30,181 @@ const Modal = ({ setShowModal, usage, product_name, questionOrProduct_id }) => {
...
     if (usage === 'addQuestion') {
       endpoint = '/qa/questions';
       body.product_id = questionOrProduct_id;
     }
 
-    postModalInput(endpoint, body);
+    try {
+      const postReq = await axios.post(localHost + endpoint, body);
+      onAorQAddition();
+        // if (usage === 'addQuestion') {
```
Comment (ryanaolea) [inline client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx]: need still?

File: server.js [hunk 9]
```diff
@@ -13,8 +13,8 @@ const headers = {
 // Middleware
 app.use(express.static(path.join(__dirname, '..', 'client/dist')));
 app.use(cors());
-app.use(express.json());
-app.use(express.urlencoded({ extended: true }));
+app.use(express.json({ limit: '50mb' }));
```
Comment (ryanaolea) [inline server.js]: this is interesting, is it for the photos? or for the text entered?

**Review by ryanaolea — CHANGES_REQUESTED** [review]
Love the logic around photo uploads. Lets talk about this more during code cleanup on wednesday. Would love to use some of this logic for modal form refactor. 
Not a lot to change except for the merge conflicts. 
Nice work on all the bug hunting!

File: client/src/components/QandA/QuestionsList.jsx (hunk 2 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList.jsx:28]: thanks

File: server.js (hunk 9 above)
Comment (FlightfulKiwi) [reply server.js]: It was, per this one tutorial I was following. Might not be helpful anymore. Removed.

File: client/src/components/QandA.jsx (hunk 1 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA.jsx]: Removed.

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx (hunk 4 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:34]: Think I've removed these.

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx (hunk 5 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:56]: Please keep for now. Will address by my next PR.

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx (hunk 6 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:84]: Thanks but painful in other respects. Leveraged Yasin massively for this.

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx (hunk 7 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx:101]: Possibly but chiefly for learning. Will change by next PR.

File: client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx (hunk 8 above)
Comment (FlightfulKiwi) [reply client/src/components/QandA/QuestionsList/QuestionCard/Modal.jsx]: Removed in latest push.

**Comment by FlightfulKiwi** [comment]
Believe I've addressed all your comments. Please note that I did remove the CSS for the file input loader button per our convo.

**Review by ryanaolea — APPROVED** [review]

**Pull request merged**